### PR TITLE
 dmd.conf modified to specify the RPATH and fixed missing sys link of libphobos 

### PR DESCRIPTION
**Pull request by ghost**
The explanation of this change is contained in the commit description

### CODE REVIEW

**Comment by WebDrake** [comment]
@ErnyTech sorry for taking a while to get to this.  Can you clean up the PR so that only the wanted changes are in here?  It's a  bit strange having a `revert` patch in a PR ;-)

File: snap/snapcraft.yaml [hunk 1]
```diff
@@ -90,6 +90,15 @@ parts:
...
     - -src
+    override-build: |
+      snapcraftctl build
+      if [ -d "../install/linux/lib64" ]; then
+        cd ../install/linux/lib64
+      else 
+        cd ../install/linux/lib32
+      fi
+      SO_FILE=$(find * -name "libphobos2.so.[0-9]*.[0-9]*.[0-9]*")
+      ln -sf $SO_FILE "${SO_FILE%.*}"
```
Comment (WebDrake) [inline snap/snapcraft.yaml]: Question: why isn't this symlink being created properly by `make install`?  Or is it being created properly but not preserved in the snap package?

I don't object to taking this change as a workaround but I'd like to confirm if we're not doing something hacky when the problem might actually be in how we're managing the install stage from the makefile.

Comment (WebDrake) [reply snap/snapcraft.yaml]: On my install I see:

```
libphobos2.a
libphobos2.so -> libphobos2.so.0.85.0 (symlink)
libphobos2.so.0.85.0
```

... which matches what is in the deb package available for download from dlang.org.  So what exactly is missing and why is it relevant/needed if upstream does not think it worth creating?

Comment (ghost) [reply snap/snapcraft.yaml]: These files are present in the deb package on dlang.org:
```
 libphobos2.so -> libphobos2.so.0.85.0
 libphobos2.so.0.85 -> libphobos2.so.0.85.0
 libphobos2.so.0.85.0`
```
We need libphobos2.so.0.85 because it is the soname of the shared library, I have no idea why make install does not create this symbolic link but Archlinux also uses another method to get pretty much the same thing (https://git.archlinux.org/svntogit/community.git/tree/trunk/PKGBUILD?h=packages/dmd#n126)

With this patch we would get this:
```
libphobos2.a
libphobos2.so -> libphobos2.so.0.85.0
libphobos2.so.0.85.0
```

to
```
libphobos2.a
libphobos2.so -> libphobos2.so.0.85.0
libphobos2.so.0.85 -> libphobos2.so.0.85.0
libphobos2.so.0.85.0

Comment (ghost) [reply snap/snapcraft.yaml]: dlang.org also seems to do the same thing (https://github.com/dlang/installer/blob/482840bf5bd8e94ebb12aabaa7ddb0dac319be90/linux/dmd_phobos.sh#L169 and https://github.com/dlang/installer/blob/482840bf5bd8e94ebb12aabaa7ddb0dac319be90/linux/dmd_rpm.sh#L191)

File: dmd-config/dmd.conf [hunk 2]
```diff
@@ -1,5 +1,5 @@
 [Environment32]
-DFLAGS=-I%@P%/../import/druntime -I%@P%/../import/phobos -L-L%@P%/../lib32 -L--export-dynamic
+DFLAGS=-I%@P%/../import/druntime -I%@P%/../import/phobos -L-L%@P%/../lib32 -L--export-dynamic -L-rpath=/snap/dmd/current/lib32 -L--enable-new-dtags
```
Comment (joseph-wakeling-frequenz) [inline dmd-config/dmd.conf]: Can you explain what the `--enable-new-dtags` flag does and why it needs to be added here?  (Not just for me, but in the commit message as well.)

File: snap/snapcraft.yaml [hunk 3]
```diff
@@ -90,6 +90,15 @@ parts:
     stage:
     - -linux
     - -src
+    override-build: |
+      snapcraftctl build
+      if [ -d "../install/linux/lib64" ]; then
+        cd ../install/linux/lib64
+      else 
+        cd ../install/linux/lib32
+      fi
```
Comment (WebDrake) [inline snap/snapcraft.yaml]: There should probably be 2 completely independent `if` checks for `lib32` and `lib64` (i.e. don't assume it's one or the other, assume that it could be either or both).  That way, if we update the package to include both 32-bit and 64-bit libs, it will continue to work without needing any modification.

File: dmd-config/dmd.conf (hunk 2 above)
Comment (WebDrake) [inline dmd-config/dmd.conf:2]: Can you explain the reason why the `--enable-new-dtags` flag is useful/needed here?

If it's for a reason other than supporting the `rpath`, it should probably be done in a separate patch.  Either way there should be a clear explanation in the commit message of why this is done.

File: snap/snapcraft.yaml (hunk 3 above)
Comment (WebDrake) [reply snap/snapcraft.yaml]: Also, probably best to avoid `cd` and instead just use the path when creating the symlink (if that makes sense).

Finally: should this really be an `override-build` or would it be better as an `override-install`?  That might avoid the need to peek into the higher-level dir with `..`.

```
override-install: |
  snapcraftctl install
  # create the symlink
```

**Comment by WebDrake** [comment]
I asked in the D release engineering slack channel why the symlink is added during package builds but not by `make install`.  It sounds like this was done just because the change was easier to make there, which is maybe understandable as a short-term workaround, but really not so good for long term maintenance.  So we should probably pressure upstream to address that at some point (perhaps in the next release).

Anyway, it sounds reasonable that in the short term at least we should indeed manually add this symlink.  I've added some comments above for how to make this a bit cleaner and more future proof -- do these sound reasonable to you?

File: dmd-config/dmd.conf (hunk 2 above)
Comment (ghost) [reply dmd-config/dmd.conf:2]: I think I have explained it in this commit https://github.com/dlang-snaps/dmd.snap/pull/31/commits/6b94eeb1b147a9d374c4cd40d414c49b45b7416c

File: snap/snapcraft.yaml (hunk 3 above)
Comment (ghost) [reply snap/snapcraft.yaml]: I don't think there is an override-install

File: dmd-config/dmd.conf (hunk 2 above)
Comment (WebDrake) [reply dmd-config/dmd.conf:2]: Ah, sorry, missed that.  Looks good.

File: snap/snapcraft.yaml (hunk 3 above)
Comment (WebDrake) [reply snap/snapcraft.yaml]: > I don't think there is an override-install

Ah, no worries.  There is an `install`, but I can't remember if that's deprecated or not.  In any case, can you update to have two separate `if` checks, so that we don't have to worry about whether the package gets updated to contain both `lib64` and `lib32` ... ?

Comment (ghost) [reply snap/snapcraft.yaml]: unfortunately, install has been deprecated

Comment (ghost) [reply snap/snapcraft.yaml]: I modified the scriptlet, now I think it looks good

**Comment by ghost** [comment]
I think too that we should implement the creation of the link in the DMD build system, it is really inconvenient for all those who create packages

File: snap/snapcraft.yaml [hunk 4]
```diff
@@ -90,6 +90,18 @@ parts:
     stage:
     - -linux
     - -src
+    override-build: |
+      snapcraftctl build
+      SO_FILE_32=$(find ../install/linux/lib32/* -name "libphobos2.so.[0-9]*.[0-9]*.[0-9]*")
```
Comment (WebDrake) [inline snap/snapcraft.yaml]: I'm afraid this fails for me when I run in `snapcraft cleanbuild`:

```
mkdir -p ../install/src/phobos/etc
mkdir -p ../install/src/phobos/std
cp -r std/* ../install/src/phobos/std/
cp -r etc/* ../install/src/phobos/etc/
cp LICENSE_1_0.txt ../install/phobos-LICENSE.txt
make[1]: Leaving directory '/root/build_dmd/parts/phobos/build'
find: ‘../install/linux/lib32/*’: No such file or directory
Failed to run 'override-build': Exit code was 1.
```

Comment (ghost) [reply snap/snapcraft.yaml]: oh, ok now it works

File: snap/snapcraft.yaml [hunk 5]
```diff
@@ -90,6 +90,12 @@ parts:
     stage:
     - -linux
     - -src
+    override-build: |
+      snapcraftctl build
+      [ -d ../install/linux/lib32 ] && SO_FILE_32=$(find ../install/linux/lib32/* -name "libphobos2.so.[0-9]*.[0-9]*.[0-9]*")
+      [ -d ../install/linux/lib64 ] && SO_FILE_64=$(find ../install/linux/lib64/* -name "libphobos2.so.[0-9]*.[0-9]*.[0-9]*")
+      [ -d ../install/linux/lib32 ] && [ -f $SO_FILE_32 ] && ln -sf $SO_FILE_32 ${SO_FILE_32%.*}
+      [ -d ../install/linux/lib64 ] && [ -f $SO_FILE_64 ] && ln -sf $SO_FILE_64 ${SO_FILE_64%.*}
```
Comment (WebDrake) [inline snap/snapcraft.yaml]: This is still not working, I'm afraid. :-(

Here is what I currently have in the existing dmd 2.085.1 snap package:

```
-rw-r--r-- 1 root root 69M Apr 10 19:41 libphobos2.a
lrwxrwxrwx 1 root root  20 Apr 10 19:41 libphobos2.so -> libphobos2.so.0.85.1
-rwxr-xr-x 1 root root 11M Apr 10 19:42 libphobos2.so.0.85.1
```

Here's what I get when I install the snap package built from this PR:

```
-rw-r--r-- 1 root root 68M Apr 17 13:58 libphobos2.a
lrwxrwxrwx 1 root root  20 Apr 17 13:58 libphobos2.so -> libphobos2.so.0.85.0
lrwxrwxrwx 1 root root  43 Apr 17 13:58 libphobos2.so.0.85 -> ../install/linux/lib64/libphobos2.so.0.85.0
-rwxr-xr-x 1 root root 11M Apr 17 13:59 libphobos2.so.0.85.0
```

Unsurprisingly, the second symlink is broken.  I suspect you need to use the `-r` option with `ln` to make sure the links are specified relative to the link location.

i.e. use

```
ln -sfr
```

rather than just `ln -sf` ;-)

Comment (WebDrake) [reply snap/snapcraft.yaml]: I tested `ln -sfr` locally and it results in:

```
-rw-r--r-- 1 root root 68M Apr 17 17:51 libphobos2.a
lrwxrwxrwx 1 root root  20 Apr 17 17:51 libphobos2.so -> libphobos2.so.0.85.0
lrwxrwxrwx 1 root root  20 Apr 17 17:51 libphobos2.so.0.85 -> libphobos2.so.0.85.0
-rwxr-xr-x 1 root root 11M Apr 17 17:52 libphobos2.so.0.85.0
```

... which I think is what is wanted, no ... ?

**Comment by WebDrake** [comment]
Besides the fix requested, I think this now needs a rebase on the latest 2.085 branch ;-)

**Review by WebDrake — COMMENTED** [review]
LGTM!  Before I give final approval and take this in, can I ask you to do one quick last re-check that a package built from this works for the use-case that motivated it (drepl, IIRC ... ?)

**Comment by ghost** [comment]
Seems to solve the problem correctly

Test build of drepl with patch applied (dmd 2.085.1)
```
erny@workstation:~/drepl$ dub build --force
Performing "debug" build using dmd for x86_64.
colorize 1.0.5: building configuration "library"...
stdx-allocator 2.77.2: building configuration "library"...
libdparse 0.8.7: building configuration "library"...
drepl 0.2.1+commit.2.g18203c0: building configuration "console"...
Running pre-build commands...
Linking...
```

Running drepl with patch applied (dmd 2.085.1)
```
erny@workstation:~/drepl$ ./drepl
Welcome to D REPL.
D> import std.stdio;
std
D> writeln("Hello D");
Hello D

D> 

```

Libraries linked with patch applied (dmd 2.085.1)
```
erny@workstation:~/drepl$ ldd drepl
	linux-vdso.so.1 (0x00007ffd6c898000)
	libgtk3-nocsd.so.0 => /usr/lib/x86_64-linux-gnu/libgtk3-nocsd.so.0 (0x00007fa5c176b000)
	libphobos2.so.0.85 => /snap/dmd/current/lib64/libphobos2.so.0.85 (0x00007fa5c0ccb000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007fa5c0cc5000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007fa5c0ca4000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007fa5c0b17000)
	librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007fa5c0b0d000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007fa5c0af1000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fa5c0907000)
	/lib64/ld-linux-x86-64.so.2 (0x00007fa5c1cfe000)

```

Test build of drepl without patch applied (dmd 2.085.1)
```
erny@workstation:~/drepl$ dub build --force
Performing "debug" build using dmd for x86_64.
colorize 1.0.5: building configuration "library"...
stdx-allocator 2.77.2: building configuration "library"...
libdparse 0.8.7: building configuration "library"...
drepl 0.2.1+commit.2.g18203c0: building configuration "console"...
Running pre-build commands...
Linking...
```

Running drepl without patch applied (dmd 2.085.1)
```
erny@workstation:~/drepl$ ./drepl
./drepl: error while loading shared libraries: libphobos2.so.0.85: cannot open shared object file: No such file or directory
```

Libraries linked without patch applied (dmd 2.085.1)
```
erny@workstation:~/drepl$ ldd drepl
	linux-vdso.so.1 (0x00007ffd29974000)
	libgtk3-nocsd.so.0 => /usr/lib/x86_64-linux-gnu/libgtk3-nocsd.so.0 (0x00007f4459926000)
	libphobos2.so.0.85 => not found
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f4459920000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f44598ff000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f4459772000)
	librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007f4459768000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f445974c000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f4459562000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f4459eb9000)
```

**Review by WebDrake — APPROVED** [review]
Great!  Let's get this in.  Thanks for fixing this, it's something I would likely not have spotted.
